README.md
# @polkadot/schnorrkel-js

A fork of [@parity/schnorrkel-js](https://github.com/paritytech/schnorrkel-js) that allows proper operation against all the environments that the `@polkadot/api` supports. Changes from the base repo -

- WASM initialisation is done async, via Promise (this allows for clear operation in webpack environments without additional workers)
- TextDecoder is polyfilled by using the version from `@polkadot/util` (consistent support, even on mobile)
- WASM outputs are optimised via `wasm-opt` from the [binaryen](https://github.com/WebAssembly/binaryen) project
- Output bundle is wrapped with camelCase names (not including the `__***` internal functions)
- Full named functions and parameter TypeScript definitions
- Requires for crypto is wrapped, removing "on-demand-require" warnings in webpack environment
- WASM output is done via a base-64 encoded string, supporting both Node.js and browser environments
- Extensive code cleanups and addition of functions required for all sr25519 operations
- Extended tests to cover Rust, wasm (via Node) and wasm (via jest), remove (here unused) www interfaces
- Attempt at asm.js support (not active, commented in build.sh)

## development

1. Build can be done via `./build.sh`
2. Tests can be done via `./test.sh`


CHANGELOG.md
# 0.2.0

- Add soft and hard key derivation functions
- Cleanup all interfaces and follow best wasm-pack practices
- Rename exposed functions to camelCase

# 0.1.0

- Initial version


test.sh
#!/usr/bin/env bash

./build.sh

echo "*** Running cargo tests"
cargo test --release -- --nocapture

echo "*** Running wasm tests"
node ./test/wasm.js

echo "*** Running jest tests"
yarn jest ./test/jest.spec.js


test/all.js
// @ts-check
const crypto = require('crypto');
const { assert, hexToU8a, stringToU8a, u8aToHex } = require('@polkadot/util');
const schnorrkel = require('../pkg/index');

function extractKeys (pair) {
  const pk = pair.slice(64);
  const sk = pair.slice(0, 64);

  return [pair, pk, sk];
}

function randomPair () {
  const seed = crypto.randomBytes(32);
  const pair = schnorrkel.keypairFromSeed(seed);

  assert(pair.length === 96, 'ERROR: Invalid pair created');

  return extractKeys(pair);
}

async function beforeAll () {
  return schnorrkel.waitReady();
}

async function pairFromSeed () {
  const SEED = stringToU8a('12345678901234567890123456789012');
  const PAIR = hexToU8a(
    '0x' +
    'f0106660c3dda23f16daa9ac5b811b963077f5bc0af89f85804f0de8e424f050' +
    'f98d66f39442506ff947fd911f18c7a7a5da639a63e8d3b4e233f74143d951c1' +
    '741c08a06f41c596608f6774259bd9043304adfa5d3eea62760bd9be97634d63'
  );

  const pair = schnorrkel.keypairFromSeed(SEED);

  console.log('\tSEC', u8aToHex(pair.slice(0, 64)));
  console.log('\tPUB', u8aToHex(pair.slice(64)));

  assert(u8aToHex(pair) === u8aToHex(PAIR), 'ERROR: pairFromSeed() does not match');
}

async function devFromSeed () {
  const SEED = hexToU8a('0xfac7959dbfe72f052e5a0c3c8d6530f202b02fd8f9f5ca3580ec8deb7797479e');
  const PAIR = hexToU8a(
    '0x' +
    '28b0ae221c6bb06856b287f60d7ea0d98552ea5a16db16956849aa371db3eb51' +
    'fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34' +
    '46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a'
  );

  const pair = schnorrkel.keypairFromSeed(SEED);

  console.log('\tSEC', u8aToHex(pair.slice(0, 64)));
  console.log('\tPUB', u8aToHex(pair.slice(64)));

  assert(u8aToHex(pair) === u8aToHex(PAIR), 'ERROR: devFromSeed() does not match');
}

async function verifyExisting () {
  const PK = hexToU8a('0x741c08a06f41c596608f6774259bd9043304adfa5d3eea62760bd9be97634d63');
  const MESSAGE = stringToU8a('this is a message');
  const SIGNATURE = hexToU8a(
    '0x' +
    'decef12cf20443e7c7a9d406c237e90bcfcf145860722622f92ebfd5eb4b5b39' +
    '90b6443934b5cba8f925a0ae75b3a77d35b8490cbb358dd850806e58eaf72904'
  );

  const isValid = schnorrkel.verify(SIGNATURE, MESSAGE, PK);

  console.log('\tRES', isValid);

  assert(isValid, 'ERROR: Unable to verify signature');
}

async function signAndVerify () {
  const MESSAGE = stringToU8a('this is a message');

  const [, pk, sk] = randomPair();
  const signature = schnorrkel.sign(pk, sk, MESSAGE);
  const isValid = schnorrkel.verify(signature, MESSAGE, pk);

  console.log('\tSIG', u8aToHex(signature));
  console.log('\tRES', isValid);

  assert(isValid, 'ERROR: Unable to verify signature');
}

async function deriveHard () {
  const CC = hexToU8a('0x0c666f6f00000000000000000000000000000000000000000000000000000000');
  const [pair] = randomPair();
  const derived = schnorrkel.deriveKeypairHard(pair, CC);

  console.log('\tSEC', u8aToHex(derived.slice(0, 64)));
  console.log('\tPUB', u8aToHex(derived.slice(64)));
}

async function deriveHardKnown () {
  const CC = hexToU8a('0x14416c6963650000000000000000000000000000000000000000000000000000');
  const PAIR = hexToU8a(
    '0x' +
    '28b0ae221c6bb06856b287f60d7ea0d98552ea5a16db16956849aa371db3eb51' +
    'fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34' +
    '46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a'
  );
  const PUBLIC = '0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d';

  const derived = schnorrkel.deriveKeypairHard(PAIR, CC);
  const publicKey = u8aToHex(derived.slice(64));

  console.log('\tSEC', u8aToHex(derived.slice(0, 64)));
  console.log('\tPUB', publicKey);

  assert(publicKey === PUBLIC, 'Unmatched resulting public keys');
}

async function deriveSoft () {
  const CC = hexToU8a('0x0c666f6f00000000000000000000000000000000000000000000000000000000');

  const [pair] = randomPair();
  const derived = schnorrkel.deriveKeypairSoft(pair, CC);

  console.log('\tSEC', u8aToHex(derived.slice(0, 64)));
  console.log('\tPUB', u8aToHex(derived.slice(64)));
}

async function deriveSoftKnown () {
  const CC = hexToU8a('0x0c666f6f00000000000000000000000000000000000000000000000000000000');
  const PUBLIC = '0x40b9675df90efa6069ff623b0fdfcf706cd47ca7452a5056c7ad58194d23440a';
  const PAIR = hexToU8a(
    '0x' +
    '28b0ae221c6bb06856b287f60d7ea0d98552ea5a16db16956849aa371db3eb51' +
    'fd190cce74df356432b410bd64682309d6dedb27c76845daf388557cbac3ca34' +
    '46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a'
  );

  const derived = schnorrkel.deriveKeypairSoft(PAIR, CC);
  const publicKey = u8aToHex(derived.slice(64));

  console.log('\tSEC', u8aToHex(derived.slice(0, 64)));
  console.log('\tPUB', publicKey);

  assert(publicKey === PUBLIC, 'Unmatched resulting public keys');
}

async function deriveSoftPubkey () {
  const CC = hexToU8a('0x0c666f6f00000000000000000000000000000000000000000000000000000000');
  const PUBLIC = hexToU8a('0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a');
  const DERIVED = '0x40b9675df90efa6069ff623b0fdfcf706cd47ca7452a5056c7ad58194d23440a';

  const derived = u8aToHex(schnorrkel.derivePublicSoft(PUBLIC, CC));

  console.log('\tPUB', derived);

  assert(derived === DERIVED, 'Unmatched resulting public keys');
}

async function benchmark () {
  const MESSAGE = stringToU8a('this is a message');

  for (let i = 0; i < 256; i++) {
    const [, pk, sk] = randomPair();

    const signature = schnorrkel.sign(pk, sk, MESSAGE);
    const isValid = schnorrkel.verify(signature, MESSAGE, pk);

    assert(isValid, 'ERROR: Unable to verify signature');
  }
}

const tests = {
  pairFromSeed,
  devFromSeed,
  deriveHard,
  deriveHardKnown,
  deriveSoft,
  deriveSoftKnown,
  deriveSoftPubkey,
  signAndVerify,
  verifyExisting,
  benchmark
};

async function runAll () {
  Object.keys(tests).forEach(async (name) => {
    console.time(name);
    console.log();
    console.log(name);

    await tests[name]();

    console.timeEnd(name);
  });
}

module.exports.beforeAll = beforeAll;
module.exports.runAll = runAll;
module.exports.tests = tests;


test/asm.js
// @ts-check
global.WebAssembly = null;

require('./wasm');


build.sh
#!/usr/bin/env bash

# install wasm-pack as required
if ! [ -x "$(command -v wasm-pack)" ]; then
  echo "*** Installing wasm-pack"
  curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
fi

# install binaryen as required
if [ ! -d "binaryen" ]; then
  echo "*** Installing binaryen"
  git clone --recursive https://github.com/WebAssembly/binaryen.git
  rm -rf binaryen/test
fi

# build wasm2js as required
# if [ ! -f "binaryen/bin/wasm2js" ]; then
#   echo "*** Building wasm2js"
#   cd binaryen
#   cmake .
#   make wasm2js
#   cd ..
# fi

# build wasm-opt as required
if [ ! -f "binaryen/bin/wasm-opt" ]; then
  echo "*** Building wasm-opt"
  cd binaryen
  cmake .
  make wasm-opt
  cd ..
fi

# install deps
echo "*** Installing JS dependencies"
yarn

# cleanup old
echo "*** Claning old builds"
rm -rf ./pkg

# build new via nightly & wasm-pack
echo "*** Building WASM output"
rustup run nightly wasm-pack build --release --scope polkadot --target nodejs

# optimise
echo "*** Optimising WASM output"
binaryen/bin/wasm-opt pkg/schnorrkel_js_bg.wasm -Os -o pkg/schnorrkel_js_opt.wasm

# build asmjs version from the input (optimised) WASM
# echo "*** Building asm.js version"
# binaryen/bin/wasm2js --no-validation --output pkg/schnorrkel_js_temp.js pkg/schnorrkel_js_opt.wasm

# convert wasm to base64 structure
echo "*** Packing WASM into base64"
./pack-node.sh

# copy our package interfaces
echo "*** Copying package sources"
cp src/js/* pkg/

# shortcuts for files
echo "*** Adjusting output"
BGJ=pkg/schnorrkel_js_bg.js
SRC=pkg/schnorrkel_js.js
DEF=pkg/schnorrkel_js.d.ts
TMP=pkg/schnorrkel_js_temp.js
ASM=pkg/schnorrkel_js_asm.js
PKG=pkg/package.json

# update the files (new addition)
# excluded: "schnorrkel_js_asm\.js",
sed -i -e 's/schnorrkel_js_bg\.wasm/schnorrkel_js_wasm\.js", "index\.js", "index\.d\.ts", "crypto-polyfill\.js/g' $PKG
sed -i -e 's/"main": "schnorrkel_js\.js"/"main": "index\.js"/g' $PKG
sed -i -e 's/"types": "schnorrkel_js\.d\.ts"/"types": "index\.d\.ts"/g' $PKG

# cleanup asm
# sed -i -e 's/import {/\/\/ import {/g' $TMP
# sed -i -e 's/function asmFunc/var schnorrkel = require('\''\.\/schnorrkel_js'\''); function asmFunc/g' $TMP
# sed -i -e 's/export const /module\.exports\./g' $TMP
# sed -i -e 's/{abort.*},memasmFunc/schnorrkel, memasmFunc/g' $TMP

# we do not want the __ imports (used by WASM) to clutter up
sed -i -e 's/var wasm;/const crypto = require('\''crypto'\''); let wasm; const requires = { crypto };/g' $SRC

# this creates issues in both the browser and RN (@polkadot/util has a polyfill)
sed -i -e 's/const TextDecoder = require('\''util'\'')\.TextDecoder;/const { u8aToString } = require('\''@polkadot\/util'\'');/g' $SRC

# TextDecoder is not available on RN, so use the @polkadot/util replacement (with polyfill)
sed -i -e 's/let cachedTextDecoder = new /\/\/ let cachedTextDecoder = new /g' $SRC
sed -i -e 's/cachedTextDecoder\.decode/u8aToString/g' $SRC

# we are swapping to a async interface for webpack support (wasm limits)
sed -i -e 's/wasm = require/\/\/ wasm = require/g' $SRC

# pull the requires from the imports and the `requires` object
sed -i -e 's/return addHeapObject(require(varg0));/return addHeapObject(requires[varg0]);/g' $SRC
# sed -i -e 's/return addHeapObject(require(varg0));/throw new Error(`Invalid require from WASM for ${varg0}`);/g' $SRC

# construct our promise and add ready helpers
echo "
module.exports.abort = function () { throw new Error('abort'); };

const createPromise = require('./schnorrkel_js_bg');
const wasmPromise = createPromise().catch(() => null);

module.exports.isReady = function () { return !!wasm; }
module.exports.waitReady = function () { return wasmPromise.then(() => !!wasm); }

wasmPromise.then((_wasm) => { wasm = _wasm });
" >> $SRC

# add extra methods to type definitions
echo "
export function isReady(): boolean;
export function waitReady(): Promise<boolean>;
" >> $DEF

# create the init promise handler
echo "
const asm = null; // require('./schnorrkel_js_asm');
const bytes = require('./schnorrkel_js_wasm');
const schnorrkel = require('./schnorrkel_js');

module.exports = async function createExportPromise () {
  const imports = {
    './schnorrkel_js': schnorrkel
  };

  if (!WebAssembly) {
    return asm;
  }

  try {
    const { instance } = await WebAssembly.instantiate(bytes, imports);

    return instance.exports;
  } catch (error) {
    return asm;
  }
}
" > $BGJ

# cleanup in-place sed files
echo "*** Cleaning up in-place edits"
rm -rf pkg/*-e

# pass through uglify to make the actual output smaller
# echo "*** Optimising asm.js output"
# yarn run uglifyjs $TMP --compress --mangle --timings --output $ASM
# mv $TMP $ASM


pre-publish.sh
#!/usr/bin/env bash

# install wasm-pack as required
if ! [ -x "$(command -v wasm-pack)" ]; then
  echo "*** Installing wasm-pack"
  curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
fi

# install binaryen as required
if [ ! -d "binaryen" ]; then
  echo "*** Installing binaryen"
  git clone --recursive https://github.com/WebAssembly/binaryen.git
  rm -rf binaryen/test
fi

# build wasm2js as required
# if [ ! -f "binaryen/bin/wasm2js" ]; then
#   echo "*** Building wasm2js"
#   cd binaryen
#   cmake .
#   make wasm2js
#   cd ..
# fi

# build wasm-opt as required
if [ ! -f "binaryen/bin/wasm-opt" ]; then
  echo "*** Building wasm-opt"
  cd binaryen
  cmake .
  make wasm-opt
  cd ..
fi

# install deps
echo "*** Installing JS dependencies"
yarn

# cleanup old
echo "*** Claning old builds"
rm -rf ./pkg

# build new via nightly & wasm-pack
echo "*** Building WASM output"
rustup run nightly wasm-pack build --release --scope polkadot --target nodejs

# optimise
echo "*** Optimising WASM output"
binaryen/bin/wasm-opt pkg/schnorrkel_js_bg.wasm -Os -o pkg/schnorrkel_js_opt.wasm

# build asmjs version from the input (optimised) WASM
# echo "*** Building asm.js version"
# binaryen/bin/wasm2js --no-validation --output pkg/schnorrkel_js_temp.js pkg/schnorrkel_js_opt.wasm

# convert wasm to base64 structure
echo "*** Packing WASM into base64"
./pack-node.sh

# copy our package interfaces
echo "*** Copying package sources"
cp src/js/* pkg/

# shortcuts for files
echo "*** Adjusting output"
BGJ=pkg/schnorrkel_js_bg.js
SRC=pkg/schnorrkel_js.js
DEF=pkg/schnorrkel_js.d.ts
TMP=pkg/schnorrkel_js_temp.js
ASM=pkg/schnorrkel_js_asm.js
PKG=pkg/package.json

# update the files (new addition)
# excluded: "schnorrkel_js_asm\.js",
sed -i -e 's/schnorrkel_js_bg\.wasm/schnorrkel_js_wasm\.js", "index\.js", "index\.d\.ts/g' $PKG
sed -i -e 's/"main": "schnorrkel_js\.js"/"main": "index\.js"/g' $PKG
sed -i -e 's/"types": "schnorrkel_js\.d\.ts"/"types": "index\.d\.ts"/g' $PKG

# cleanup asm
# sed -i -e 's/import {/\/\/ import {/g' $TMP
# sed -i -e 's/function asmFunc/var schnorrkel = require('\''\.\/schnorrkel_js'\''); function asmFunc/g' $TMP
# sed -i -e 's/export const /module\.exports\./g' $TMP
# sed -i -e 's/{abort.*},memasmFunc/schnorrkel, memasmFunc/g' $TMP

# we do not want the __ imports (used by WASM) to clutter up
sed -i -e 's/var wasm;/const crypto = require('\''crypto'\''); let wasm; const requires = { crypto };/g' $SRC

# this creates issues in both the browser and RN (@polkadot/util has a polyfill)
sed -i -e 's/const TextDecoder = require('\''util'\'')\.TextDecoder;/const { u8aToString } = require('\''@polkadot\/util'\'');/g' $SRC

# TextDecoder is not available on RN, so use the @polkadot/util replacement (with polyfill)
sed -i -e 's/let cachedTextDecoder = new /\/\/ let cachedTextDecoder = new /g' $SRC
sed -i -e 's/cachedTextDecoder\.decode/u8aToString/g' $SRC

# we are swapping to a async interface for webpack support (wasm limits)
sed -i -e 's/wasm = require/\/\/ wasm = require/g' $SRC

# pull the requires from the imports and the `requires` object
sed -i -e 's/return addHeapObject(require(varg0));/return addHeapObject(requires[varg0]);/g' $SRC
# sed -i -e 's/return addHeapObject(require(varg0));/throw new Error(`Invalid require from WASM for ${varg0}`);/g' $SRC

# construct our promise and add ready helpers
echo "
module.exports.abort = function () { throw new Error('abort'); };

const createPromise = require('./schnorrkel_js_bg');
const wasmPromise = createPromise().catch(() => null);

module.exports.isReady = function () { return !!wasm; }
module.exports.waitReady = function () { return wasmPromise.then(() => !!wasm); }

wasmPromise.then((_wasm) => { wasm = _wasm });
" >> $SRC

# add extra methods to type definitions
echo "
export function isReady(): boolean;
export function waitReady(): Promise<boolean>;
" >> $DEF

# create the init promise handler
echo "
const asm = null; // require('./schnorrkel_js_asm');
const bytes = require('./schnorrkel_js_wasm');
const schnorrkel = require('./schnorrkel_js');

module.exports = async function createExportPromise () {
  const imports = {
    './schnorrkel_js': schnorrkel
  };

  if (!WebAssembly) {
    return asm;
  }

  try {
    const { instance } = await WebAssembly.instantiate(bytes, imports);

    return instance.exports;
  } catch (error) {
    return asm;
  }
}
" > $BGJ

# cleanup in-place sed files
echo "*** Cleaning up in-place edits"
rm -rf pkg/*-e

# pass through uglify to make the actual output smaller
# echo "*** Optimising asm.js output"
# yarn run uglifyjs $TMP --compress --mangle --timings --output $ASM
# mv $TMP $ASM


src/wrapper.rs
use schnorrkel::{signing_context, Keypair, SecretKey, MiniSecretKey, PublicKey,
	derive::{Derivation, ChainCode, CHAIN_CODE_LENGTH},
	keys::{KEYPAIR_LENGTH, MINI_SECRET_KEY_LENGTH, PUBLIC_KEY_LENGTH, SECRET_KEY_LENGTH},
	sign::{Signature, SIGNATURE_LENGTH}
};

// We must make sure that this is the same as declared in the substrate source code.
const SIGNING_CTX: &[u8] = b"substrate";
const DERIVE_CTX: &[u8] = b"SchnorrRistrettoHDKD";

/// Private helper function.
fn keypair_from_seed(seed: &[u8]) -> Keypair {
	MiniSecretKey::from_bytes(seed)
		.expect("32 bytes can always build a key; qed")
		.expand_to_keypair()
}

pub fn __hard_derive_keypair(pair: &[u8], cc: &[u8]) -> [u8; KEYPAIR_LENGTH] {
	let pair = match Keypair::from_bytes(pair) {
		Ok(pair) => pair,
		Err(_) => panic!("Provided pair is invalid.")
	};
	let derived = pair
		.secret
		.hard_derive_mini_secret_key(signing_context(DERIVE_CTX).bytes(&cc[..]))
		.expand_to_keypair()
		.to_bytes();

	let mut res = [0u8; KEYPAIR_LENGTH];
	res.copy_from_slice(&derived);
	res
}

pub fn __hard_derive_seed(seed: &[u8], cc: &[u8]) -> [u8; MINI_SECRET_KEY_LENGTH] {
	let derived = keypair_from_seed(seed)
		.secret
		.hard_derive_mini_secret_key(signing_context(DERIVE_CTX).bytes(&cc[..]))
		.to_bytes();

	let mut res = [0u8; MINI_SECRET_KEY_LENGTH];
	res.copy_from_slice(&derived);
	res
}

pub fn __hard_derive_secret(secret: &[u8], cc: &[u8]) -> [u8; SECRET_KEY_LENGTH] {
	let secret = match SecretKey::from_bytes(secret) {
		Ok(secret) => secret,
		Err(_) => panic!("Provided private key is invalid.")
	};
	let derived = secret
		.hard_derive_mini_secret_key(signing_context(DERIVE_CTX).bytes(&cc[..]))
		.expand();

	let mut res = [0u8; SECRET_KEY_LENGTH];
	res.copy_from_slice(&derived.to_bytes());
	res
}

pub fn __keypair_from_seed(seed: &[u8]) -> [u8; KEYPAIR_LENGTH] {
	let keypair = keypair_from_seed(seed).to_bytes();
	let mut kp = [0u8; KEYPAIR_LENGTH];
	kp.copy_from_slice(&keypair);
	kp
}

pub fn __soft_derive_keypair(pair: &[u8], junction: &[u8]) -> [u8; KEYPAIR_LENGTH] {
	let keypair = match Keypair::from_bytes(pair) {
		Ok(keypair) => keypair,
		Err(_) => panic!("Provided keypair is invalid.")
	};
	let mut cc = [0u8; CHAIN_CODE_LENGTH];
	cc.copy_from_slice(&junction);
	let derived = keypair.derived_key_simple(ChainCode(cc), &[]).0;
	let mut res = [0u8; KEYPAIR_LENGTH];
	res.copy_from_slice(&derived.to_bytes());
	res
}

pub fn __soft_derive_public(pubkey: &[u8], junction: &[u8]) -> [u8; PUBLIC_KEY_LENGTH] {
	let public = match PublicKey::from_bytes(pubkey) {
		Ok(public) => public,
		Err(_) => panic!("Provided public key is invalid.")
	};
	let mut cc = [0u8; CHAIN_CODE_LENGTH];
	cc.copy_from_slice(&junction);
	let derived = public.derived_key_simple(ChainCode(cc), &[]).0;
	let mut res = [0u8; PUBLIC_KEY_LENGTH];
	res.copy_from_slice(&derived.to_bytes());
	res
}

pub fn __soft_derive_secret(seckey: &[u8], junction: &[u8]) -> [u8; SECRET_KEY_LENGTH] {
	let secret = match SecretKey::from_bytes(seckey) {
		Ok(secret) => secret,
		Err(_) => panic!("Provided private key is invalid.")
	};
	let mut cc = [0u8; CHAIN_CODE_LENGTH];
	cc.copy_from_slice(&junction);
	let derived = secret.derived_key_simple(ChainCode(cc), &[]).0;
	let mut res = [0u8; SECRET_KEY_LENGTH];
	res.copy_from_slice(&derived.to_bytes());
	res
}

pub fn __secret_from_seed(seed: &[u8]) -> [u8; SECRET_KEY_LENGTH] {
	let secret = keypair_from_seed(seed).secret.to_bytes();
	let mut s = [0u8; SECRET_KEY_LENGTH];
	s.copy_from_slice(&secret);
	s
}

pub fn __verify(signature: &[u8], message: &[u8], pubkey: &[u8]) -> bool {
	let sig = match Signature::from_bytes(signature) {
		Ok(some_sig) => some_sig,
		Err(_) => return false
	};
	let pk = match PublicKey::from_bytes(pubkey) {
		Ok(some_pk) => some_pk,
		Err(_) => return false
	};
	pk.verify_simple(SIGNING_CTX, message, &sig)
}

pub fn __sign(public: &[u8], private: &[u8], message: &[u8]) -> [u8; SIGNATURE_LENGTH] {
	// despite being a method of KeyPair, only the secret is used for signing.
	let secret = match SecretKey::from_bytes(private) {
		Ok(some_secret) => some_secret,
		Err(_) => panic!("Provided private key is invalid.")
	};

	let public = match PublicKey::from_bytes(public) {
		Ok(some_public) => some_public,
		Err(_) => panic!("Provided public key is invalid.")
	};

	let context = signing_context(SIGNING_CTX);
	secret.sign(context.bytes(message), &public).to_bytes()
}


src/lib.rs
extern crate wasm_bindgen;
extern crate wee_alloc;
extern crate schnorrkel;

use schnorrkel::{
	Keypair, MiniSecretKey, PublicKey, SecretKey, Signature,
	derive::{Derivation, ChainCode, CHAIN_CODE_LENGTH},
};
use wasm_bindgen::prelude::*;

// Use `wee_alloc` as the global allocator.
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

// We must make sure that this is the same as declared in the substrate source code.
const SIGNING_CTX: &'static [u8] = b"substrate";

/// ChainCode construction helper
fn create_cc(data: &[u8]) -> ChainCode {
	let mut cc = [0u8; CHAIN_CODE_LENGTH];

	cc.copy_from_slice(&data);

	ChainCode(cc)
}

/// Keypair helper function.
fn create_from_seed(seed: &[u8]) -> Keypair {
	match MiniSecretKey::from_bytes(seed) {
		Ok(mini) => return mini.expand_to_keypair(),
		Err(_) => panic!("Provided seed is invalid.")
	}
}

/// Keypair helper function.
fn create_from_pair(pair: &[u8]) -> Keypair {
	match Keypair::from_bytes(pair) {
		Ok(pair) => return pair,
		Err(_) => panic!("Provided pair is invalid.")
	}
}

/// PublicKey helper
fn create_public(public: &[u8]) -> PublicKey {
	match PublicKey::from_bytes(public) {
		Ok(public) => return public,
		Err(_) => panic!("Provided public key is invalid.")
	}
}

/// SecretKey helper
fn create_secret(secret: &[u8]) -> SecretKey {
	match SecretKey::from_bytes(secret) {
		Ok(secret) => return secret,
		Err(_) => panic!("Provided private key is invalid.")
	}
}

/// Perform a derivation on a secret
///
/// * secret: UIntArray with 64 bytes
/// * cc: UIntArray with 32 bytes
///
/// returned vector the derived keypair as a array of 96 bytes
#[wasm_bindgen]
pub fn derive_keypair_hard(pair: &[u8], cc: &[u8]) -> Vec<u8> {
	create_from_pair(pair).secret
		.hard_derive_mini_secret_key(Some(create_cc(cc)), &[]).0
		.expand_to_keypair()
		.to_bytes()
		.to_vec()
}

/// Perform a derivation on a secret
///
/// * secret: UIntArray with 64 bytes
/// * cc: UIntArray with 32 bytes
///
/// returned vector the derived keypair as a array of 96 bytes
#[wasm_bindgen]
pub fn derive_keypair_soft(pair: &[u8], cc: &[u8]) -> Vec<u8> {
	create_from_pair(pair)
		.derived_key_simple(create_cc(cc), &[]).0
		.to_bytes()
		.to_vec()
}

/// Perform a derivation on a publicKey
///
/// * pubkey: UIntArray with 32 bytes
/// * cc: UIntArray with 32 bytes
///
/// returned vector is the derived publicKey as a array of 32 bytes
#[wasm_bindgen]
pub fn derive_public_soft(public: &[u8], cc: &[u8]) -> Vec<u8> {
	create_public(public)
		.derived_key_simple(create_cc(cc), &[]).0
		.to_bytes().to_vec()
}

/// Generate a key pair.
///
/// * seed: UIntArray with 32 element
///
/// returned vector is the concatenation of first the private key (64 bytes)
/// followed by the public key (32) bytes.
#[wasm_bindgen]
pub fn keypair_from_seed(seed: &[u8]) -> Vec<u8> {
	create_from_seed(seed)
		.to_bytes()
		.to_vec()
}

/// Sign a message
///
/// The combination of both public and private key must be provided.
/// This is effectively equivalent to a keypair.
///
/// * public: UIntArray with 32 element
/// * private: UIntArray with 64 element
/// * message: Arbitrary length UIntArray
///
/// * returned vector is the signature consisting of 64 bytes.
#[wasm_bindgen]
pub fn sign(public: &[u8], secret: &[u8], message: &[u8]) -> Vec<u8> {
	create_secret(secret)
		.sign_simple(SIGNING_CTX, message, &create_public(public))
		.to_bytes()
		.to_vec()
}

/// Verify a message and its corresponding against a public key;
///
/// * signature: UIntArray with 64 element
/// * message: Arbitrary length UIntArray
/// * pubkey: UIntArray with 32 element
#[wasm_bindgen]
pub fn verify(signature: &[u8], message: &[u8], public: &[u8]) -> bool {
	let signature = match Signature::from_bytes(signature) {
		Ok(signature) => signature,
		Err(_) => return false
	};

	create_public(public)
		.verify_simple(SIGNING_CTX, message, &signature)
}

#[cfg(test)]
pub mod tests {
	extern crate rand;
	extern crate schnorrkel;

	use hex_literal::{hex, hex_impl};
	use super::*;
	use schnorrkel::{SIGNATURE_LENGTH, KEYPAIR_LENGTH, SECRET_KEY_LENGTH};

	fn generate_random_seed() -> Vec<u8> {
		(0..32).map(|_| rand::random::<u8>() ).collect()
	}

	#[test]
	fn can_create_keypair() {
		let seed = generate_random_seed();
		let keypair = keypair_from_seed(seed.as_slice());

		assert!(keypair.len() == KEYPAIR_LENGTH);
	}

	#[test]
	fn creates_pair_from_known() {
		let seed = hex!("fac7959dbfe72f052e5a0c3c8d6530f202b02fd8f9f5ca3580ec8deb7797479e");
		let expected = hex!("46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a");
		let keypair = keypair_from_seed(&seed);
		let public = &keypair[SECRET_KEY_LENGTH..KEYPAIR_LENGTH];

		assert_eq!(public, expected);
	}

	#[test]
	fn can_sign_message() {
		let seed = generate_random_seed();
		let keypair = keypair_from_seed(seed.as_slice());
		let private = &keypair[0..SECRET_KEY_LENGTH];
		let public = &keypair[SECRET_KEY_LENGTH..KEYPAIR_LENGTH];
		let message = b"this is a message";
		let signature = sign(public, private, message);

		assert!(signature.len() == SIGNATURE_LENGTH);
	}

	#[test]
	fn can_verify_message() {
		let seed = generate_random_seed();
		let keypair = keypair_from_seed(seed.as_slice());
		let private = &keypair[0..SECRET_KEY_LENGTH];
		let public = &keypair[SECRET_KEY_LENGTH..KEYPAIR_LENGTH];
		let message = b"this is a message";
		let signature = sign(public, private, message);

		assert!(verify(&signature[..], message, public));
	}

	#[test]
	fn soft_derives_pair() {
		let cc = hex!("0c666f6f00000000000000000000000000000000000000000000000000000000"); // foo
		let seed = hex!("fac7959dbfe72f052e5a0c3c8d6530f202b02fd8f9f5ca3580ec8deb7797479e");
		let expected = hex!("40b9675df90efa6069ff623b0fdfcf706cd47ca7452a5056c7ad58194d23440a");
		let keypair = keypair_from_seed(&seed);
		let derived = derive_keypair_soft(&keypair, &cc);
		let public = &derived[SECRET_KEY_LENGTH..KEYPAIR_LENGTH];

		assert_eq!(public, expected);
	}

	#[test]
	fn soft_derives_public() {
		let cc = hex!("0c666f6f00000000000000000000000000000000000000000000000000000000"); // foo
		let public = hex!("46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a");
		let expected = hex!("40b9675df90efa6069ff623b0fdfcf706cd47ca7452a5056c7ad58194d23440a");

		assert_eq!(derive_public_soft(&public, &cc), expected);
	}

	#[test]
	fn hard_derives_pair() {
		let cc = hex!("14416c6963650000000000000000000000000000000000000000000000000000"); // Alice
		let seed = hex!("fac7959dbfe72f052e5a0c3c8d6530f202b02fd8f9f5ca3580ec8deb7797479e");
		let expected = hex!("d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d");
		let keypair = keypair_from_seed(&seed);
		let derived = derive_keypair_hard(&keypair, &cc);
		let public = &derived[SECRET_KEY_LENGTH..KEYPAIR_LENGTH];

		assert_eq!(public, expected);
	}
}
